File: 1547-minimum-cost-to-cut-a-stick/1547-minimum-cost-to-cut-a-stick.cpp
class Solution {
public:
    int minCost(int n, vector<int>& cuts) {
        cuts.push_back(0);
        cuts.push_back(n);
        sort(cuts.begin(), cuts.end());

        int m = cuts.size();
        vector<vector<int>> dp(m, vector<int>(m, 0));

        for (int len = 2; len < m; len++) {
            for (int start = 0; start < m - len; start++) {
                int end = start + len;
                int minCost = INT_MAX;

                for (int k = start + 1; k < end; k++) {
                    int cost = cuts[end] - cuts[start];
                    int totalCost = dp[start][k] + dp[k][end];
                    minCost = min(minCost, totalCost + cost);
                }

                dp[start][end] = minCost;
            }
        }

        return dp[0][m - 1];
    }
};

File: 1547-minimum-cost-to-cut-a-stick/1547-minimum-cost-to-cut-a-stick.java
class Solution {
    static int f(int i, int j, int[] c, int[][] dp) {
        if(i > j) return 0;
        if(dp[i][j] != -1) return dp[i][j];
        
        int mini = Integer.MAX_VALUE;
        for(int k = i; k <= j; ++k) {
            int cost = (c[j+1]-c[i-1]) + f(i, k-1, c, dp) + f(k+1, j, c, dp);
            mini = Math.min(mini, cost);
        }
        return dp[i][j] = mini;
    }
    public int minCost(int n, int[] cuts) {
        int x = cuts.length;
        int[][] dp = new int[x+1][x+1];
        for(int[] it : dp) Arrays.fill(it, -1);
        int[] c = new int[x+2];
        c[0] = 0;
        c[x+1] = n;
        for(int i = 0; i < x; i++) c[i+1] = cuts[i];
        Arrays.sort(c);
        return f(1, x, c, dp);
    }
}